README.md
# NODEJS AZAMPAY SDK

AzamPay NodeJs SDK to help you easily and seamlessly integrate with Azampay APIs.

## Table of Contents

1. [Installations](#installation-and-use)
2. [Use](#use) <br>
   2.1 [Get Token](#get-token) <br>
   2.2 [Bank Checkout](#bank-checkout) <br>
   2.3 [Mno Checkout](#mno-checkout) <br>
   2.4 [Disbursement](#disbursement) <br>
   2.5 [Name Lookup](#name-lookup) <br>
   2.6 [Transaction Status](#transaction-status) <br>
   2.7 [Post Checkout](#post-checkout) <br>
   2.7 [Payment Partners](#payment-partners)

<br>

# Installation and Use

From terminal in the root directory of your project, run

```sh

npm i azampay

```

<br>

# Use

<details>
<summary>ES6 Import</summary>

```JS

import azampay from 'azampay'

```

</details>

<details>
<summary>Module Require</summary>

```JS

const azampay = require('azampay').default

```

</details>

- The `import` or `require` from `azampay` package exports two properties.

  1.  `getToken` method that is used to retrieve access token from Azam Pay
  2.  `instance` a class that can be used or extended to add functionalities to the `azampay` package

## Get Token

`Requirement Definition`

| Property     | Description                                                          | Type   | Required |
| ------------ | -------------------------------------------------------------------- | ------ | -------- |
| env          | Enum: SANDBOX \| LIVE. Azampay environment. Default SANDBOX          | String | [ ]      |
| clientId     | It will be the client id generated during application registration.  | String | [x]      |
| appName      | It will be the name of application.                                  | String | [x]      |
| clientSecret | It will be the secret key generated during application registration. | String | [x]      |
| apiKey       | Azam Pay API key given as token in the settings page.                | String | [x]      |

`Request Payload`

```JSON

{
  "env": "string",
  "clientId": "string",
  "appName": "string",
  "clientSecret": "string",
  "apiKey": "string",
}

```

`Request Method`

```JS

const token = await azampay.getToken(payload);

```

`Response`

If successful, the response will be of type [TokenResponse](#token-response) or [ErrorResponse](#error-response)

```JS

{
  data: { accessToken: string; expire: string };
  success: boolean;
  message: string;
  code: string | number;
  statusCode: number;
  bankCheckout: (
    payload: BankCheckout,
    options: RequestOptions
  ) => Promise<CheckoutResponse>;

  mnoCheckout: (
    payload: MnoCheckout,
    options: RequestOptions
  ) => Promise<CheckoutResponse>;

  postCheckout: (
    payload: PostCheckOut,
    options: RequestOptions
  ) => Promise<PostCheckOutInterface>;

  disburse: (
    payload: Disburse,
    options: RequestOptions
  ) => Promise<DisburseResponse>;

  transactionStatus: (
    payload: TransactionStatus,
    options: RequestOptions
  ) => Promise<TransactionStatusResponse>;

  nameLookup: (
    payload: NameLookup,
    options: RequestOptions
  ) => Promise<NameLookupResponse>;
}

```

`Response Definition`

| Property          | Description                                                                              | Type             |
| ----------------- | ---------------------------------------------------------------------------------------- | ---------------- |
| data              | Azam Pay respnse with access token and expire time                                       | Object           |
| success           | A `true` boolean value indicating that the request was successfull                       | Boolean          |
| message           | Response message                                                                         | String           |
| code              | Response code.                                                                           | Number \| String |
| statusCode        | Response Http Status code. Possibly `200` or `201`                                       | Number           |
| bankCheckout      | A method to initiate a Bank checkout with Azam Pay.                                      | Method           |
| mnoCheckout       | A method to initiate MNO checkout with Azam Pay.                                         | Method           |
| disburse          | A method used to lookup the name associated with a bank account or Mobile Money account  | Method           |
| transactionStatus | A method used to retrieve the status of a disbursement transaction made through AzamPay. | Method           |
| nameLookup        | A method used to lookup the name associated with a bank account or Mobile Money account. | Method           |

<br>

# Bank Checkout

Initiating a bank checkout, we can use two methods, one from the token results method and the other from the Azam Pay instance exported from the azampay package.

Bank checkout method takes two arguments, the first with the request payload for Azam Pay bank checkout and second is [optional with additional options](#request-options).

`Request Payload`

```JS

{

  amount: string;
  currencyCode: string;
  merchantAccountNumber: string;
  merchantMobileNumber: string;
  merchantName?: string | null;
  otp: string;
  provider: 'CRDB' | 'NMB' |
  referenceId: string;
  additionalProperties?: Record<string, unknown> | null;
}

```

1. `Request Payload Definition`

| Property              | Definition                                                                                                 | Type   |
| --------------------- | ---------------------------------------------------------------------------------------------------------- | ------ |
| amount                | This is amount that will be charged from the given account.                                                | String |
| currencyCode          | Code of currency                                                                                           | String |
| merchantAccountNumber | This is the account number/MSISDN that consumer will provide. The amount will be deducted from this.       | String |
| merchantMobileNumber  | Mobile number                                                                                              | String |
| merchantName          | Nullable consumer name.                                                                                    | String |
| otp                   | One time password                                                                                          | String |
| provider              | Enum: CRDB \| NMB BankProvider.                                                                            | String |
| referenceId           | This id belongs to the calling application. Maximum Allowed length for this field is 128 ascii characters. | String |
| additionaProperties   | This is additional JSON data that calling application can provide. This is optional.                       | Object |

2. `Request Options Definition`

- These are optional if you use the method from the response in [getToken](#get-token) otherwise mandatory if using the instance and the options were not passed during instantiation.

`Method`

<details>
<summary>Request from Token Response</summary>

```JS

await token.bankCheckout(payload, options)

```

</details>

<details>
<summary>Request from Instance</summary>

```JS

const instance = new azampay.instance({accessToken: token.data.accessToken,apiKey: 'YOUR API KEY'})
await instance.bankCheckout(payload, options)

```

</details>

`Response`

```JS
{
  transactionId: string,
  message: string,
  succcess: true
}

```

#

NB: Every response has a `success` property denoting whether the request was successful or not and if it wasn't, the response will be of type [Error Response](#error-response)

<br>

## MNO Checkout

For request method explanation, refer to [Bank Checkout](#bank-checkout)'s explanation.

1. `Request Payload Definition`

| Property            | Definition                                                                                                   | Type   |
| ------------------- | ------------------------------------------------------------------------------------------------------------ | ------ |
| amount              | This is amount that will be charged from the given account.                                                  | String |
| currencyCode        | Code of currency                                                                                             | String |
| accountNumber       | This is the account number/MSISDN that consumer will provide. The amount will be deducted from this account. | String |
| provider            | Enum: Airtel \| Tigo \| Halopesa \| Azampesa \| Mpesa.                                                       | String |
| externalId          | This id belongs to the calling application. Maximum Allowed length for this field is 128 ascii characters.   | String |
| additionaProperties | This is additional JSON data that calling application can provide. This is optional.                         | Object |

```JS

{
  accountNumber: string;
  amount: string;
  currency: string;
  externalId: string;
  provider: 'Airtel' | 'Tigo' | 'Halopesa' | 'Azampesa' | 'Mpesa' | string;
  additionalProperties?: Record<string, unknown> | null;
}

```

`Method`

<details>
<summary>Request from Token Response</summary>

```JS

await token.mnoCheckout(payload, options)

```

</details>

<details>
<summary>Request from Instance</summary>

```JS

const instance = new azampay.instance({accessToken: token.data.accessToken,apiKey: 'YOUR API KEY'})
await instance.mnoCheckout(payload, options)

```

</details>

`Response`
The response is the same as the [Bank Checkout](#bank-checkout)'s

<br>

## Disbursement

This method allows for the transfer of money from other countries to Tanzania. It requires the authorization token generated above, passed as a header in the request. The request should also contain details of the source, destination, and transfer details. Additionally, the request can include an external reference ID and remarks.

This method takes two arguments, mandatory payload and the other optional options as explained in [Bank Checkout](#bank-checkout)

`Payload`

```JS

{
  source: Source;
  destination: Destination;
  transferDetails: TransferDetails;
  externalReferenceId: string;
  remarks: string;
}

```

| Property            | Description                                         | Type                             |
| ------------------- | --------------------------------------------------- | -------------------------------- |
| source              | Contains information about the source account.      | [Object](#source-or-destination) |
| destination         | Contains information about the destination account. | [Object](#source-or-destination) |
| transferDetails     | Contains information about the transfer.            | [Object](#transfer-details)      |
| externalReferenceId | An external reference ID to track the transaction.  | String                           |
| remarks             | Any remarks to be included with the transaction.    | String                           |

`Method`

<details>
<summary>Request from Token Response</summary>

```JS

await token.disburse(payload, options)

```

</details>

<details>
<summary>Request from Instance</summary>

```JS

const instance = new azampay.instance({accessToken: token.data.accessToken,apiKey: 'YOUR API KEY'})
await instance.disburse(payload, options)

```

</details>

`Response`

```JS

{
  data: string;
  message: string;
  success: boolean;
  statusCode: number;
}

```

`Description`

| Property   | Description                                                           | Type    |
| ---------- | --------------------------------------------------------------------- | ------- |
| data       | A string containing the status of the transaction.                    | String  |
| message    | A string containing a human-readable message describing the response. | String  |
| success    | A boolean indicating whether the request was successful or not.       | Boolean |
| statusCode | An integer indicating the status code of the response.                | Number  |

<br>

## Name Lookup

This method is used to lookup the name associated with a bank account or Mobile Money account.

This method takes two arguments, mandatory payload and the other optional options as explained in [Bank Checkout](#bank-checkout)

`Payload`

```JS

{
  bankName: string;
  accountNumber: string;
}

```

`Description`

| Property      | Description                                                 | Type   |
| ------------- | ----------------------------------------------------------- | ------ |
| bankName      | Bank name or Mobile Money name associated with the account. | String |
| accountNumber | Bank account number or Mobile Money number.                 | String |

`Method`

<details>
<summary>Request from Token Response</summary>

```JS

await token.nameLookup(payload, options)

```

</details>

<details>
<summary>Request from Instance</summary>

```JS

const instance = new azampay.instance({accessToken: token.data.accessToken,apiKey: 'YOUR API KEY'})
await instance.nameLookup(payload, options)

```

</details>

`Response`

```JS

{
  name: string;
  message: string;
  success: boolean;
  accountNumber: string;
  bankName: string;
}

```

`Description`

| Property      | Description                                                      | Type    |
| ------------- | ---------------------------------------------------------------- | ------- |
| bankName      | Bank name or Mobile Money name associated with the account.      | String  |
| accountNumber | Bank account number or Mobile Money number.                      | String  |
| name          | Name associated with the account.                                | String  |
| message       | A brief description of the response status.                      | String  |
| success       | A boolean value indicating if the request was successful or not. | Boolean |

<br>

## Transaction Status

This method allows you to retrieve the status of a disbursement transaction made through AzamPay.
This method takes two arguments, mandatory payload and the other optional options as explained in [Bank Checkout](#bank-checkout)

`Payload`

```JS

{
  reference: string;
  bankName: string;
}

```

`Description`

| Property  | Description                                                                              | Type   |
| --------- | ---------------------------------------------------------------------------------------- | ------ |
| reference | The transaction ID you received when making the disbursement request.                    | String |
| bankName  | The name of the mobile network operator (MNO) you used to make the disbursement request. | String |

`Method`

<details>
<summary>Request from Token Response</summary>

```JS

await token.transactionStatus(payload, options)

```

</details>

<details>
<summary>Request from Instance</summary>

```JS

const instance = new azampay.instance({accessToken: token.data.accessToken,apiKey: 'YOUR API KEY'})
await instance.transactionStatus(payload, options)

```

</details>

`Response`

| Property   | Description                                                           | Type    |
| ---------- | --------------------------------------------------------------------- | ------- |
| data       | A string containing the status of the transaction.                    | String  |
| message    | A string containing a human-readable message describing the response. | String  |
| success    | A boolean indicating whether the request was successful or not.       | Boolean |
| statusCode | An integer indicating the status code of the response.                | Number  |

<br>

## Post Checkout

For this post request, send all params that are mentioned below to this end point.

This end point will respond back with the URL of your payments. Merchant Application can open this url in a new window to continue with the checkout process of the transaction

<br>

`Payload`

```JS

{
  appName: string;
  clientId: string;
  vendorId: string;
  language: string;
  currency: string;
  externalId: string;
  requestOrigin: string;
  redirectFailURL: string;
  redirectSuccessURL: string;
  vendorName: string;
  amount: string;
  cart: Cart;
}

```

`Description`

| Property           | Description                                                 | Type   |
| ------------------ | ----------------------------------------------------------- | ------ |
| amount             | This is amount that will be charged from the given account. | String |
| appName            | This is the application name.                               | String |
| cart               | Shoping cart with multiple item.                            | Object |
| clientId           | Client id is unique id for identify client.                 | String |
| externalId         | 30 charecters long unique string.                           | String |
| language           | Language code for transalate the application.               | String |
| redirectFailURL    | URL that you want to redirected to at transaction failure.  | String |
| redirectSuccessURL | URL that you want to redirected to at transaction success.  | String |
| requestOrigin      | URL which the request is being originated.                  | String |
| vendorId           | Unique id for validate vendor.                              | String |
| vendorName         | Name of vendor.                                             | String |

`Response`

This returns a success boolean property indicating whether the operation was successful or not and a data string

```JS

{
  data: string
  success: boolean;
  [key: string]: unknown;
}

```

<br>

## Payment Partners

This method will return the registered partners of the provided merchant

This method takes optional options argument as explained in [Bank Checkout](#bank-checkout)

`Method`

<details>
<summary>Request from Token Response</summary>

```JS

await token.partners(options)

```

</details>

<details>
<summary>Request from Instance</summary>

```JS

const instance = new azampay.instance({accessToken: token.data.accessToken,apiKey: 'YOUR API KEY'})
await instance.partners(options)

```

</details>

`Response`

```JS
{
  success: boolean;
  partners: Partners[];
}

```

`Description`


CHANGELOG.md
# AZAMPAY NODEJS SDK

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

[0.0.0] - 2023-05-18

Added

- Release first major version of the SDK

[0.0.1] - 2023-05-19

Added

- Partners method to get payment partners

Changed

- Improve README to include new methods

[0.0.2] - 2023-05-23

Added

- Add status code to all successful requests

Changed

- Move all interfaces to shared base interface for import in external packages

[0.0.3] - 2023-06-06

Changed

- Change urls for live

[0.0.4] - 2023-06-07

Added

- Improvement in error response messages


src/shared/interfaces/base.interface.ts
/**
 * @interface TokenPayload Request payload for token.
 */
export interface TokenPayload {
  /**
   * @string It will be the name of application.
   */
  appName: string;

  /**
   * @string It will be the client id generated during application registration.
   */
  clientId: string;

  /**
   * @string It will be the secret key generated during application registration.
   */
  clientSecret: string;

  /**
   * @string Azam Pay API key given as token in the settings page.
   */
  apiKey: string;

  /**
   * @string [SANDBOX | LIVE] Azam Pay environment.
   */
  env?: 'SANDBOX' | 'LIVE';
}

/**
 * @interface BankCheckout Bank checkout and make payment to requested provider.
 */
export interface BankCheckout {
  /**
   * @string This is amount that will be charged from the given account.
   */
  amount: string;

  /**
   * @string Code of currency
   */
  currencyCode: 'TZS' | string;

  /**
   * @string This is the account number/MSISDN that consumer will provide. The amount will be deducted from this account.
   */
  merchantAccountNumber: string;

  /**
   * @string Mobile number
   */
  merchantMobileNumber: string;

  /**
   * @string Nullable consumer name.
   */
  merchantName?: string | null;

  /**
   * @string One time password
   */
  otp: string;

  /**
   * @string [CRDB | NMB] BankProvider.
   */
  provider: 'CRDB' | 'NMB' | string;

  /**
   * @string This id belongs to the calling application. Maximum Allowed length for this field is 128 ascii characters.
   */
  referenceId: string;

  /**
   * @Object This is additional JSON data that calling application can provide. This is optional.
   */
  additionalProperties?: Record<string, unknown> | null;
}

/**
 * @interface MnoCheckout Mobile Money Operator checkout and make payment to requested provider.
 */
export interface MnoCheckout {
  /**
   * @string This is the account number/MSISDN that consumer will provide. The amount will be deducted from this account.
   */
  accountNumber: string;

  /**
   *@string This is amount that will be charged from the given account.
   */
  amount: string;

  /**
   * @string This is the transaciton currency. Current support values are only (TZS)
   */
  currency: string;

  /**
   * @string This id belongs to the calling application. Maximum Allowed length for this field is 128 ascii characters
   */
  externalId: string;

  /**
   * @string ['Airtel' | 'Tigo' | 'Halopesa' | 'Azampesa' | 'Mpesa'] MNO Provider.
   */
  provider: 'Airtel' | 'Tigo' | 'Halopesa' | 'Azampesa' | 'Mpesa' | string;

  /**
   * @object This is additional JSON data that calling application can provide. This is optional.
   */
  additionalProperties?: Record<string, unknown> | null;
}

/**
 * @interface CheckoutResponse Response for both BankCheckout and MnoCheckout
 */
export interface CheckoutResponse {
  /**
   * @string Each successfull transaction will be given a valid transaction id.
   */
  transactionId: string;

  /**
   * @string This is the status message of checkout request.
   */
  msg: string;

  /**
   * @string This is the status message of checkout request.
   */
  message: string;

  /**
   * @boolean A boolean value indicating if the request was successful or not.
   */
  success: boolean;

  /**
   * @number Response Http Status code.
   */
  statusCode: number;
}

/**
 * Request options for request method
 */
export interface RequestOptions {
  /**
   * Token from Azampay app registration
   */
  apiKey?: string;

  /**
   * Access token from Azampay got on request
   */
  accessToken?: string;

  /**
   * Azampay environment
   */
  env?: 'LIVE' | 'SANDBOX';
}

/**
 * @interface PostCheckOut Request interface
 */
export interface PostCheckOut {
  /**
   * @string It will be the name of application.
   */
  appName: string;

  /**
   * @string Client id is unique id for identify client
   */
  clientId: string;

  /**
   * @string [uuid] Unique id for validate vendor
   */
  vendorId: string;

  /**
   * @string Language code for transalate the application
   */
  language: string;

  /**
   *@string Currency code that will convert amount into specific currency format
   */
  currency: string;

  /**
   * @string 30 charecters long unique string
   */
  externalId: string;

  /**
   * @string The URL where the request is originating
   */
  requestOrigin: string;

  /**
   * @string  URL that you want to redirected to at transaction failure
   */
  redirectFailURL: string;

  /**
   * @string URL that you want to redirected to at transaction success
   */
  redirectSuccessURL: string;

  /**
   * @string Name of vendor
   */
  vendorName: string;

  /**
   * @string This is amount that will be charged from the given account.
   */
  amount: string;

  /**
   * @object Shoping cart with multiple item
   */
  cart: Cart;
}

export interface Disburse {
  /**
   * @object Contains information about the source account.
   */
  source: Source;

  /**
   *@object Contains information about the destination account.
   */
  destination: Destination;

  /**
   *@object Contains information about the transfer.
   */
  transferDetails: TransferDetails;

  /**
   * @string An external reference ID to track the transaction.
   */
  externalReferenceId: string;

  /**
   * @string Any remarks to be included with the transaction.
   */
  remarks: string;
}

/**
 * @interface Source Account Details
 */
export interface Source {
  /**
   * @string Country code
   */
  countryCode: string;

  /**
   * @string Source account full name
   */
  fullName: string;

  /**
   * @string Source account bank name
   */
  bankName: string;

  /**
   * @string Source account account number
   */
  accountNumber: string;

  /**
   * @string Source account currency
   */
  currency: string;
}

/**
 * @interface Destination Account Details
 */
export interface Destination {
  /**
   * @string Country code
   */
  countryCode: string;

  /**
   * @string Destination account full name
   */
  fullName: string;

  /**
   * @string Destination account bank name
   */
  bankName: string;

  /**
   * @string Destination account account number
   */
  accountNumber: string;

  /**
   * @string Destination account currency
   */
  currency: string;
}

/**
 * @interface TransferDetails
 */
export interface TransferDetails {
  /**
   * @string Transfer type
   */
  type: string;

  /**
   * @number Transfer amount
   */
  amount: number;

  /**
   * @string Transfer date
   */
  date: string;
}

/**
 * @interface DisburseResponse Response for disbursement
 */

export interface DisburseResponse {
  /**
   * @string A string containing the status of the transaction.
   */
  data: string;

  /**
   * @string A string containing a human-readable message describing the response.
   */
  message: string;

  /**
   * @boolean A boolean indicating whether the request was successful or not.
   */
  success: boolean;

  /**
   * @number Http status code for the request
   */
  statusCode: number;
}

/**
 * @interface NameLookup An interface for looking up account name
 */
export interface NameLookup {
  /**
   * @string Bank name or Mobile Money name associated with the account.
   */
  bankName: string;

  /**
   * @string Bank account number or Mobile Money number.
   */
  accountNumber: string;
}

/**
 * @interface NameLookupResponse An interface used to lookup the name associated with a bank account or Mobile Money account.
 */
export interface NameLookupResponse {
  /**
   * @string Name associated with the account.
   */
  name: string;

  /**
   * @string A brief description of the response status.
   */
  message: string;

  /**
   * @boolean A boolean value indicating if the request was successful or not.
   */
  success: boolean;

  /**
   * @string Account number.
   */
  accountNumber: string;

  /**
   * @string Bank name or Mobile Money name associated with the account.
   */
  bankName: string;

  /**
   * @number Response Http Status code.
   */
  statusCode: number;
}

/**
 * @interface TransactionStatusResponse This API allows you to retrieve the status of a disbursement transaction made through AzamPay.
 */
export interface TransactionStatusResponse {
  /**
   * @string A string containing the status of the transaction.
   */
  data: string;

  /**
   * @string A string containing a human-readable message describing the response.
   */
  message: string;

  /**
   * @boolean A boolean indicating whether the request was successful or not.
   */
  success: boolean;

  /**
   * @number An integer indicating the status code of the response.
   */
  statusCode: number;
}

/**
 * @interface TransactionStatus Interface for status method that allows you to retrieve the status of a disbursement transaction made through AzamPay.
 */
export interface TransactionStatus {
  /**
   *@string The transaction ID you received when making the disbursement request.
   */
  reference: string;

  /**
   * @string The name of the mobile network operator (MNO) you used to make the disbursement request.
   */
  bankName: string;
}

/**
 * @interface TokenResponse Token request response with bankCheckout and mnoCheckout.
 */
export interface TokenResponse {
  /**
   * @interface TOKENDETAILS Token request response with access token and expiry time.
   */
  data?: TOKENDETAILS;

  /**
   * @string This is the status message of checkout request.
   */
  message: unknown;

  /**
   * @boolean A boolean value indicating if the request was successful or not.
   */
  success: boolean;

  /**
   * @number Response Http Status code.
   */
  statusCode: number;

  /**
   * @method bankCheckout A method to initiate a Bank checkout with Azam Pay.
   */
  bankCheckout?: (
    payload: BankCheckout,
    options?: RequestOptions
  ) => Promise<CheckoutResponse>;

  /**
   * @method mnoCheckout A method to initiate MNO checkout with Azam Pay.
   */
  mnoCheckout?: (
    payload: MnoCheckout,
    options?: RequestOptions
  ) => Promise<CheckoutResponse>;

  /**
   * @method postCheckout A method to initiate post checkout with Azam Pay.
   */
  postCheckout?: (
    payload: PostCheckOut,
    options?: RequestOptions
  ) => Promise<PostCheckOutInterface>;

  /**
   * @method disburse A method used to lookup the name associated with a bank account or Mobile Money account.
   */

  disburse?: (
    payload: Disburse,
    options?: RequestOptions
  ) => Promise<DisburseResponse>;

  /**
   * @method transactionStatus Allows you to retrieve the status of a disbursement transaction made through AzamPay.
   */

  transactionStatus?: (
    payload: TransactionStatus,
    options?: RequestOptions
  ) => Promise<TransactionStatusResponse>;

  /**
   * @method nameLookup Is used to lookup the name associated with a bank account or Mobile Money account.
   */

  nameLookup?: (
    payload: NameLookup,
    options?: RequestOptions
  ) => Promise<NameLookupResponse>;

  /**
   * @method partners A method to retrieve payment partner
   */

  partners: (
    options?: RequestOptions
  ) => Promise<PartnersResponse | ErrorResponse>;
}

/**
 * @interface Partners Payment partners
 */
export interface Partners {
  /**
   * @string Payment Partner logo URL
   */
  logoUrl: string;

  /***
   * @string Name of the payment partner e.g (Azampesa, Airtel, Halopesa, Tigopesa, vodacom)
   */
  partnerName: string;

  /**
   * @string Provider enum value e.g (airtel=2, tigo=3, halopesa=4, azampesa=5, Mpesa=10)
   */
  provider: number;

  /**
   * @string Name of the vendor
   */
  vendorName: string;

  /**
   * @string Unique id for payment vendor
   */
  paymentVendorId: string;

  /**
   * @string Unique id for payment partner
   */
  paymentPartnerId: string;

  /**
   * @string Currency code that will convert amount into specific currency format
   */
  currency: string;
}

/**
 * @interface PartnersResponse
 */
export interface PartnersResponse {
  /**
   * @number Response Http Status code.
   */
  statusCode: number;

  /**
   * @boolean A boolean value indicating if the request was successful or not.
   */
  success: boolean;

  /**
   * @interface Partners An array of payment partners
   */
  partners: Partners[];
}

/**
 * @interface AzamPayToken Token request response with access token details and status code.
 */
export interface AzamPayToken {
  /**
   * @interface TOKENDETAILS Token request response with access token and expiry time.
   */
  data: TOKENDETAILS;

  /**
   * @string Response message.
   */
  message: string;

  /**
   * @boolean A boolean value indicating if the request was successful or not.
   */
  success: boolean;

  /**
   * @number Response Http Status code.
   */
  statusCode: number;
}

/**
 * @interface TOKENDETAILS Token request response with access token and expiry time.
 */
export interface TOKENDETAILS {
  /**
   * @string Azam Pay access token.
   */
  accessToken: string;

  /**
   * @string Azam Pay access token expire time.
   */
  expire: string;
}

/**
 * @interface ErrorResponse Response when an error occurred while doing a request to Azam Pay APIs
 */

export interface ErrorResponse {
  /**
   * @boolean A boolean value indicating if the request was successful or not.
   */
  success: boolean;

  /**
   * @string Response message.
   */
  message: string;

  /**
   * @string Response code.
   */
  code?: string;

  /**
   * @number Response Http Status code.
   */
  statusCode: number;

  /**
   * @Object Additional response object.
   */
  [key: string]: unknown;

  data?: TOKENDETAILS;
  bankCheckout?: (
    payload: BankCheckout,
    options?: RequestOptions
  ) => Promise<CheckoutResponse>;
  mnoCheckout?: (
    payload: MnoCheckout,
    options?: RequestOptions
  ) => Promise<CheckoutResponse>;

  postCheckout?: (
    payload: PostCheckOut,
    options?: RequestOptions
  ) => Promise<string>;

  disburse?: (
    payload: Disburse,
    options?: RequestOptions
  ) => Promise<DisburseResponse>;

  transactionStatus?: (
    payload: TransactionStatus,
    options?: RequestOptions
  ) => Promise<TransactionStatusResponse>;

  nameLookup?: (
    payload: NameLookup,
    options?: RequestOptions
  ) => Promise<NameLookupResponse>;
  transactionId?: string;

  partners?: (options?: RequestOptions) => Promise<PartnersResponse[]>;
}
export interface AzamPayErrors {
  [key: string]: unknown;
}

/**
 * @interface AzamPayInstance An optional Azam Pay interface for initiating the request instance.
 */
export interface AzamPayInstance {
  /**
   * @string Azam Pay access token.
   */
  accessToken?: string;

  /**
   * @string AzamPayInstance Azam Pay API key given as token in the settings page.
   */
  apiKey?: string;

  /**
   * @string AzamPayInstance Azam Pay environment. Either SANDBOX | LIVE the default is SANDBOX.
   */
  env?: 'LIVE' | 'SANDBOX';
}

/**
 * @method PostCallback request payload
 */
export interface PostCallback {
  /**
   * @string This is the application name.
   */
  appName: string;

  /**
   * @string Client id is unique id for identify client.
   */
  clientId: string;

  /**
   * @string <uuid> Unique id for validate vendor.
   */
  vendorId: string;

  /**
   * @string Language code for transalate the application.
   */
  language: string;

  /**
   * @string Currency code that will convert amount into specific currency format.
   */
  currency: string;

  /**
   * @string 30 charecters long unique string.
   */
  externalId: string;

  /**
   * @string URL which the request is being originated.
   */
  requestOrigin: string;

  /**
   * @string URL that you want to redirected to at transaction failure.
   */
  redirectFailURL: string;

  /**
   * @string URL that you want to redirected to at transaction success.
   */
  redirectSuccessURL: string;
  /**
   * @string Name of vendor.
   */
  vendorName: string;
  /**
   * @string This is amount that will be charged from the given account.
   */
  amount: string;
  /**
   * @object Shoping cart with multiple item
   */
  cart: Cart;
}

/**
 * @object Shoping cart with multiple item
 */
export interface Cart {
  /**
   * @string An array of vendor items.
   */
  items: Item[];
}

export interface Item {
  /**
   * @string Vendor Item name.
   */
  name: string;
}

/**
 * @interface PostCheckOutInterface Response for post checkout
 */

export interface PostCheckOutInterface {
  /**
   * @string This is the status message of checkout request.
   */
  data: string;

  /**
   * @boolean A boolean value indicating if the request was successful or not.
   */
  success: boolean;

  /**
   * @number Response Http Status code.
   */
  statusCode: number;

  /**
   * @uknown Generic response for post checkout
   */
  [key: string]: unknown;
}
